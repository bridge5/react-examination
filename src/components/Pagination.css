:local(.pagination) {
  padding: 10px 0;
  display: flex;
  justify-content: center;
  list-style-type: none;
}

:local(.page) {
  padding: 0 10px;
  cursor: pointer;
}

:local(.current) {
  color: white;
  cursor: default;
}
