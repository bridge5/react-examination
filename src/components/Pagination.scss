.pagination-wrapper {
  border-top: 1px solid #f2f2f5;
  margin-top: 20px;
  ul {
      height: 26px;
      padding: 4px 0;
      line-height: 26px;
      text-align: center;
      li {
          display: inline-block;
          a {
              display: inline-block;
              color: #333;
          }
          &.page-pre,
          &.page-after {
              width: 100px;
              a {
                  cursor: pointer;
                  padding: 0 10px;
                  &.off {
                      cursor: default;
                      color: #aaa;
                  }
                  &:hover {
                      background: #f2f2f5;
                      &.off {
                          background: none;
                      }
                  }
              }
          }
          &.page-count {
              box-sizing: border-box;
              border-left: 1px solid #f2f2f5;
              border-right: 1px solid #f2f2f5;
              border-left: 1px solid #d9d9d9;
              border-right: 1px solid #d9d9d9;
              line-height: 100%;
              a {
                  padding: 0 7px;
                  height: 20px;
                  line-height: 20px;
                  margin: 0 8px;
                  border-radius: 2px;
                  &.on {
                    //   background: #f2f2f5;
                    background-color: #FFF;
                      font-size: 20px;
                  }
                  &:hover {
                      background: #f2f2f5;
                      cursor: pointer;
                      &.disable {
                          background: none;
                          cursor: default;
                      }
                  }
              }
              &.alone {
                  border: none;
              }
          }
      }
  }
}