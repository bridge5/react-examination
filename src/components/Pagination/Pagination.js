import React from 'react';
import styles from './Pagination.css';
const { pageContainer } = styles;
export default ({ currentPageIndex, changePage, totalPageIndex }) => {
  return (
    <div className={pageContainer}>
      {currentPageIndex <= 0 ? null : (
        <button className="btn btn-default" onClick={() => changePage(-1)}>
          &laquo;
        </button>
      )}
      <button className="btn btn-default">{currentPageIndex + 1}</button>
      {currentPageIndex === totalPageIndex ? null : (
        <button className="btn btn-default" onClick={() => changePage(1)}>
          &raquo;
        </button>
      )}
    </div>
  );
};
