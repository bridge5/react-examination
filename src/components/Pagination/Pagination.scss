:local(.pageContainer) {
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 10px 0;
  width: 100%;
}
