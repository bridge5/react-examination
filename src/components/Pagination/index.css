.pagination {
  display: block;
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  color: rgba(0, 0, 0, 0.65);
  font-size: 14px;
  font-variant: tabular-nums;
  line-height: 1.5715;
  list-style: none;
  font-feature-settings: "tnum";
}

.pagination-item {
  display: inline-block!important;
  text-align: center;
  min-width: 24px;
  height: 24px;
  margin: 0;
  line-height: 22px;
  list-style: none;
  border: 1px solid #d9d9d9;
  border-radius: 2px;
  outline: 0;
  cursor: pointer;
}

.pagination-item:hover {
  color: #1890ff;
}

.pagination-item--active {
  font-weight: 500;
  background: #fff;
  border-color: #1890ff;
  color: #1890ff;
}
