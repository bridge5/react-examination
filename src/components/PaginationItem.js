import React, { PureComponent } from 'react';
// import classnames from 'classnames';
import PropTypes from 'prop-types';
import styles from './Pagination.css';

class PaginationItem extends PureComponent {
  render() {
    const { page, currentPage, gotoPage } = this.props;
    const isCurrentPage = currentPage === page;
    return (
      <li
        className={`${styles.page} ${isCurrentPage ? styles.current : ''}`}
        onClick={e => gotoPage(page)}>
        {page}
      </li>
    );
  }
}

PaginationItem.propTypes = {
  page: PropTypes.number.isRequired,
  currentPage: PropTypes.number.isRequired,
  gotoPage: PropTypes.func.isRequired,
};

export default PaginationItem;
