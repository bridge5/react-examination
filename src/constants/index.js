export const POSITIONS = ['ALL', 'SF', 'PG'];
export const DEFAULT_POSITION = POSITIONS[0];
export const DEFAULT_PAGE = 1;
export const PER_LIST_ITEMS = 5;

export const DEFAULT_PLAYERS = [
  {
    name: 'LeBron James',
    team: 'LOS ANGELES LAKERS',
    position: 'SF',
    starred: true,
  },
  {
    name: 'Kevin Duran',
    team: 'GOLDEN STATE WARRIORS',
    position: 'SF',
    starred: false,
  },
  {
    name: 'Anthony Davis',
    team: 'NEW ORLEANS PELICANS',
    position: 'PF',
    starred: false,
  },
  {
    name: 'Stephen Curry',
    team: 'GOLDEN STATE WARRIORS',
    position: 'PG',
    starred: false,
  },
  {
    name: 'James Harden',
    team: 'HOUSTON ROCKETS',
    position: 'SG',
    starred: false,
  },
  {
    name: 'Kawhi Leonard',
    team: 'TORONTO RAPTORS',
    position: 'SF',
    starred: false,
  },
]
