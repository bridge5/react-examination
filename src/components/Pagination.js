import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styles from './Pagination.css';
import PaginationItem from './PaginationItem';

class Pagination extends Component {
  componentDidMount() {
    const { currentPage } = this.props;
    this.gotoPage(currentPage);
  }

  render() {
    const { currentPage } = this.props;
    const totalPages = this.getTotalPages();

    return (
      <nav>
        <ul className={styles.pagination}>
          {totalPages.map((page, index) => 
            <PaginationItem
              key={index} 
              currentPage={currentPage}
              gotoPage={this.gotoPage}
              page={page} />
          )}
        </ul>
      </nav>
    );
  }

  gotoPage = page => {
    const { onPageChanged, perListItems } = this.props;
    const totalPages = this.getTotalPages();

    const paginationData = {
      currentPage: page,
      totalPages,
      perListItems
    };

    onPageChanged(paginationData)
  };

  getTotalPages = () => {
    const { totalItems, perListItems } = this.props
    const reminder = totalItems % perListItems;
    let pages = parseInt(totalItems / perListItems, 10);
    if (reminder) {
      pages = pages + 1;
    }
    const totalPages = Array.from({ length: pages }, (v, i) => i + 1);
    return totalPages;
  };
}

Pagination.propTypes = {
  totalItems: PropTypes.number.isRequired,
  currentPage: PropTypes.number,
  perListItems: PropTypes.number,
  onPageChanged: PropTypes.func.isRequired,
};

Pagination.defaultProps = {
  currentPage: 1,
  perListItems: 3
};

export default Pagination;
