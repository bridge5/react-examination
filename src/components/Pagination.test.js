import React from 'react';
import { shallow } from 'enzyme'
import Pagination from './Pagination';
import PaginationItem from './PaginationItem';
import { DEFAULT_PLAYERS } from '../constants/';

const setup = (moreProps = {}) => {
  const defaultPlayers = DEFAULT_PLAYERS
  const props = {
    totalItems: defaultPlayers.length,
    onPageChanged: jest.fn(),
    ...moreProps
  }

  const wrapper = shallow(<Pagination {...props} />)
  return {
    props,
    wrapper
  }
}

it('should render with totalItems and onPageChanged callback function correctly without crash', () => {
  const { wrapper } = setup()
  expect(wrapper).toMatchSnapshot()
});

it('getTotalPages() - should PaginationItem component render correct number by perListItems default props', () => {
  const { wrapper } = setup()
  const totalPages = wrapper.instance().getTotalPages().length
  expect(wrapper.find(PaginationItem)).toHaveLength(totalPages)
})

it('getTotalPages() - should PaginationItem component render correct number by perListItems props', () => {
  const moreProps1 = { perListItems: 2 }
  const pagination1 = setup(moreProps1)
  const correctPages1 = (pagination1.props.totalItems % pagination1.props.perListItems)
    ? parseInt(pagination1.props.totalItems / pagination1.props.perListItems, 10) + 1
    : parseInt(pagination1.props.totalItems / pagination1.props.perListItems, 10)
  const totalPages1 = pagination1.wrapper.instance().getTotalPages().length
  expect(totalPages1).toEqual(correctPages1)

  const moreProps2 = { perListItems: 5 }
  const pagination2 = setup(moreProps2)
  const correctPages2 = (pagination2.props.totalItems % pagination2.props.perListItems)
    ? parseInt(pagination2.props.totalItems / pagination2.props.perListItems, 10) + 1
    : parseInt(pagination2.props.totalItems / pagination2.props.perListItems, 10)
  const totalPages2 = pagination2.wrapper.instance().getTotalPages().length
  expect(totalPages2).toEqual(correctPages2)
})
