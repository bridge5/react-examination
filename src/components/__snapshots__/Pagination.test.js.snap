// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should render with totalItems and onPageChanged callback function correctly without crash 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Pagination
    currentPage={1}
    onPageChanged={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "currentPage": 1,
              "perListItems": 3,
              "totalPages": Array [
                1,
                2,
              ],
            },
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    perListItems={3}
    totalItems={6}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <ul>
        <PaginationItem
          currentPage={1}
          gotoPage={[Function]}
          page={1}
        />
        <PaginationItem
          currentPage={1}
          gotoPage={[Function]}
          page={2}
        />
      </ul>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <PaginationItem
            currentPage={1}
            gotoPage={[Function]}
            page={1}
          />,
          <PaginationItem
            currentPage={1}
            gotoPage={[Function]}
            page={2}
          />,
        ],
        "className": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "class",
          "props": Object {
            "currentPage": 1,
            "gotoPage": [Function],
            "page": 1,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "class",
          "props": Object {
            "currentPage": 1,
            "gotoPage": [Function],
            "page": 2,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "ul",
    },
    "type": "nav",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <ul>
          <PaginationItem
            currentPage={1}
            gotoPage={[Function]}
            page={1}
          />
          <PaginationItem
            currentPage={1}
            gotoPage={[Function]}
            page={2}
          />
        </ul>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <PaginationItem
              currentPage={1}
              gotoPage={[Function]}
              page={1}
            />,
            <PaginationItem
              currentPage={1}
              gotoPage={[Function]}
              page={2}
            />,
          ],
          "className": undefined,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "class",
            "props": Object {
              "currentPage": 1,
              "gotoPage": [Function],
              "page": 1,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "class",
            "props": Object {
              "currentPage": 1,
              "gotoPage": [Function],
              "page": 2,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "ul",
      },
      "type": "nav",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`should render with totalItems and onPageChanged callback function correctly without crash 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Pagination
    currentPage={1}
    onPageChanged={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "currentPage": 1,
              "perListItems": 3,
              "totalPages": Array [
                1,
                2,
              ],
            },
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    perListItems={3}
    totalItems={6}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <ul>
        <PaginationItem
          currentPage={1}
          gotoPage={[Function]}
          page={1}
        />
        <PaginationItem
          currentPage={1}
          gotoPage={[Function]}
          page={2}
        />
      </ul>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <PaginationItem
            currentPage={1}
            gotoPage={[Function]}
            page={1}
          />,
          <PaginationItem
            currentPage={1}
            gotoPage={[Function]}
            page={2}
          />,
        ],
        "className": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "class",
          "props": Object {
            "currentPage": 1,
            "gotoPage": [Function],
            "page": 1,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "class",
          "props": Object {
            "currentPage": 1,
            "gotoPage": [Function],
            "page": 2,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "ul",
    },
    "type": "nav",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <ul>
          <PaginationItem
            currentPage={1}
            gotoPage={[Function]}
            page={1}
          />
          <PaginationItem
            currentPage={1}
            gotoPage={[Function]}
            page={2}
          />
        </ul>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <PaginationItem
              currentPage={1}
              gotoPage={[Function]}
              page={1}
            />,
            <PaginationItem
              currentPage={1}
              gotoPage={[Function]}
              page={2}
            />,
          ],
          "className": undefined,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "class",
            "props": Object {
              "currentPage": 1,
              "gotoPage": [Function],
              "page": 1,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "class",
            "props": Object {
              "currentPage": 1,
              "gotoPage": [Function],
              "page": 2,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "ul",
      },
      "type": "nav",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`should render with totalItems and onPageChanged callback function without crash 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Pagination
    currentPage={1}
    onPageChanged={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "currentPage": 1,
              "perListItems": 3,
              "totalPages": Array [
                1,
                2,
              ],
            },
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    perListItems={3}
    totalItems={6}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <ul>
        <PaginationItem
          currentPage={1}
          gotoPage={[Function]}
          page={1}
        />
        <PaginationItem
          currentPage={1}
          gotoPage={[Function]}
          page={2}
        />
      </ul>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <PaginationItem
            currentPage={1}
            gotoPage={[Function]}
            page={1}
          />,
          <PaginationItem
            currentPage={1}
            gotoPage={[Function]}
            page={2}
          />,
        ],
        "className": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "class",
          "props": Object {
            "currentPage": 1,
            "gotoPage": [Function],
            "page": 1,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "class",
          "props": Object {
            "currentPage": 1,
            "gotoPage": [Function],
            "page": 2,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "ul",
    },
    "type": "nav",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <ul>
          <PaginationItem
            currentPage={1}
            gotoPage={[Function]}
            page={1}
          />
          <PaginationItem
            currentPage={1}
            gotoPage={[Function]}
            page={2}
          />
        </ul>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <PaginationItem
              currentPage={1}
              gotoPage={[Function]}
              page={1}
            />,
            <PaginationItem
              currentPage={1}
              gotoPage={[Function]}
              page={2}
            />,
          ],
          "className": undefined,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "class",
            "props": Object {
              "currentPage": 1,
              "gotoPage": [Function],
              "page": 1,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "class",
            "props": Object {
              "currentPage": 1,
              "gotoPage": [Function],
              "page": 2,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "ul",
      },
      "type": "nav",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
